driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/BatchCursorPublisher.java

```diff
@@ -0,0 +1,169 @@
+
+    AsyncReadOperation<AsyncBatchCursor<T>> asAsyncFirstReadOperation() {
+        return asAsyncReadOperation();
+    }
+
+    @Nullable
+    ClientSession getClientSession() {
+        return clientSession;
+    }
+
+    MongoOperationPublisher<T> getMongoOperationPublisher() {
+        return mongoOperationPublisher;
+    }
+
+    Operations<T> getOperations() {
+        return mongoOperationPublisher.getOperations();
+    }
+
+    MongoNamespace getNamespace() {
+        return mongoOperationPublisher.getNamespace();
+    }
+
+    ReadPreference getReadPreference() {
+        return mongoOperationPublisher.getReadPreference();
+    }
+
+    CodecRegistry getCodecRegistry() {
+        return mongoOperationPublisher.getCodecRegistry();
+    }
+
+    boolean getRetryReads() {
+        return mongoOperationPublisher.getRetryReads();
+    }
+
+    Class<T> getDocumentClass() {
+        return mongoOperationPublisher.getDocumentClass();
+    }
+
+
+    @Nullable
+    public Integer getBatchSize() {
+        return batchSize;
+    }
+
+    public Publisher<T> batchSize(final int batchSize) {
+        this.batchSize = batchSize;
+        return this;
+    }
+
+    public Publisher<T> first() {
+        return batchCursor(this::asAsyncFirstReadOperation)
+                .flatMap(batchCursor -> Mono.create(sink -> {
+                    batchCursor.setBatchSize(1);
+                    Mono.from(batchCursor.next())
+                            .doOnTerminate(batchCursor::close)
+                            .doOnError(sink::error)
+                            .doOnSuccess(results -> {
+                                if (results == null || results.isEmpty()) {
+                                    sink.success();
+                                } else {
+                                    sink.success(results.get(0));
+                                }
+                            })
+                            .subscribe();
+                }));
+    }
+
+    @Override
+    public void subscribe(final Subscriber<? super T> subscriber) {
+        batchCursor()
+                .flatMapMany(batchCursor -> {
+                    AtomicBoolean inProgress = new AtomicBoolean(false);
+                    if (batchSize != null) {
+                        batchCursor.setBatchSize(batchSize);
+                    }
+                    return Flux.create((FluxSink<T> sink) -> {
+                        sink.onRequest(value -> recurseCursor(sink, batchCursor, inProgress));
+                        recurseCursor(sink, batchCursor, inProgress);
+                    }, FluxSink.OverflowStrategy.BUFFER)
+                    .doOnCancel(batchCursor::close);
+                })
+                .subscribe(subscriber);
+    }
+
+    void recurseCursor(final FluxSink<T> sink, final BatchCursor<T> batchCursor, final AtomicBoolean inProgress){
+        if (!sink.isCancelled() && sink.requestedFromDownstream() > 0 && inProgress.compareAndSet(false, true)) {
```

Comment:
Can the sink actually be cancelled, or is it internal?  It doesn't look like any tests excercise the false branch for this.

Reply:
Added a test - the subscription to this flux means the sink is cancelled.